### PR TITLE
Path drawing

### PR DESCRIPTION
**Pull request by Marko132001**
I implemented the feature to draw nodes and connect them on the faculty layout. Also there is an onclick option that will draw a node and output to the console its designated coordinates.

### CODE REVIEW

File: client/src/components/Header.tsx
Comment (JakovTomasic) [inline client/src/components/Header.tsx:9]: src does not need to be wrapped inside {}. This is just a string, no need to include typescript execution in it. Same for width and height.

Write it as you would write a simple HTML file.

File: client/src/components/ZoomableSVG.tsx
Comment (JakovTomasic) [inline client/src/components/ZoomableSVG.tsx:7]: The name K is not meaningful. X and y can maybe be improved, but they are acceptable.

K should be renamed to scale or something similar.

File: client/src/components/Map.tsx
Comment (JakovTomasic) [inline client/src/components/Map.tsx:80]: This onClick option will be useful, but should not always be enabled. Add a boolean flag for it to the props. The click should be disabled by default (if the prop is not passed to the component).

File: client/src/components/Map.tsx
Comment (JakovTomasic) [inline client/src/components/Map.tsx:17]: This line adds a number of assumptions:
- the image is located in submaps folder
- its name ends with .svg

When passing an image path to the component via props the caller of the component should define the whole path. I see you adding these assumption in other places, too. Like in DirectionsCard.

This line should be as follows:
`.attr("xlink:href", this.props.layoutImage)`

File: client/src/components/Map.tsx
Comment (JakovTomasic) [inline client/src/components/Map.tsx:19]: Hardcoded dimensions. This worked for me:
```
.attr("width", "100%")
.attr("height", "100%")
```

**Review by JakovTomasic — CHANGES_REQUESTED** [review]
I added a few comments that should be resolved. The Map component should be made entirely responsive. You hardcoded some random absolute dimensions in a few places.

A submap SVG should be cropped around its content. These hardcoded dimensions also added top and bottom padding to the image. That messes with all dimensions and we cannot draw a graph on top of it because the coordinates are not the same.

Open the SVG in google chrome and see that it does not have any margins around it. They mustn't be added because coordinates should be reliable and exact.

If you have problems implementing this using relative dimensions you can pass SVG width and height to the maps props. These dimensions are not in pixels and can be useful only for determining SVG aspect radio and maybe some other calculations. It would be better if you could find a solution that does not need SVG dimensions but is an option if needed.

**Comment by Marko132001** [comment]
I made changes based on the code review. However, the svg responsiveness problem was resolved by hardcoding values in the viewBox attribute. Further discussion is required in order to resolve this problem dynamically for all svg images, which happen to have different aspect ratios.

**Comment by JakovTomasic** [comment]
> I made changes based on the code review. However, the svg responsiveness problem was resolved by hardcoding values in the viewBox attribute. Further discussion is required in order to resolve this problem dynamically for all svg images, which happen to have different aspect ratios.

I understand. Then we will need to use the approach I suggested before (in the first review).

By defining proper dimensions, I managed to fix all issues. Check out merge request #42. Do not merge it before removing comments that I put just to explain my changes.

**Comment by Marko132001** [comment]
@JakovTomasic FYI I added one additional commit in path_drawing_fix branch before merging with this branch.

**Comment by JakovTomasic** [comment]
> @JakovTomasic FYI I added one additional commit in path_drawing_fix branch before merging with this branch.

Thank you for the notice. The commit has a great solution. Great job!

File: client/src/pages/navigation.tsx
Comment (JakovTomasic) [inline client/src/pages/navigation.tsx:15]: Resolve this comment after all others. Now we need to implement the Navigation component (i.e. navigation page) to be compatible with the rest of the app.

Navigation will receive data from the home page using Link and to=... (https://stackoverflow.com/questions/52238637/react-router-how-to-pass-data-between-pages-in-react) - fell free to suggest if there is something better.
Navigation will use the MapNavigator interface (mock it) to create a path to draw.

The navigation needs to receive needed parameters from the home page. Prop should be `startNodeId: number, endNodeFilter: MapNodeFilter` so the MapNavigator interface can be created and called in the Navigation page. Mock everything you need to get a (mocked) NavigationDirections object.

All of this is an initialization process for the navigation page. Your navigation page should then handle the NavigationDirections object. Draw the current step's path from it. (test path being on multiple submaps - but each navigation step should be only on one submap, of course).

Remove the submapId from the prop. You will find a submapId in the NavigationDirections.

As you already know, the Map component shouldn't know about anything of this. It just receives a path to draw and a map on which to draw that path.

After this, we will have a fully functional navigation page 😃

**Review by JakovTomasic — CHANGES_REQUESTED** [review]
I added individual comments where I think improvements can be made. The most notable is the final implementation of the navigation page using other interfaces. Meaning now is the time to implement real "path drawing".

As a side note, when I open a map on the desktop full-screen the map sometimes goes below my visible screen and I need to scroll. It would be better if we have a max height for the map so the page is always non-scrollable. This is not strictly related to "path drawing" so you don't need to fix it here nor "lose" your time on that because functional path drawing is more important. I'm just mentioning it here so you can start to think about the solutions in the background if you already haven't.

The implemented code is great so far. The code for this whole navigation feature will be very good.

**Comment by JakovTomasic** [comment]
Why did you recommit everything (force pushed it)? Was it the git rebase?

I would like to discuss the rebase in the next meeting. It is a very useful and cool feature, but it messes up with pull request review comments and commit timestamps.

Interesting pros and cons. But I think the cons can be avoided and we should use it.

File: client/src/components/Map.tsx
Comment (JakovTomasic) [inline client/src/components/Map.tsx:72]: This code assumes each navigation step will be two nodes (one connection) long which is not true.

You should probably draw the lines in the above .forEach (currently .map) iteration (connect the current and the previous node).

I suggest changing a mock a bit to test navigation steps with e.g. three nodes.

File: client/src/components/Map.tsx
Comment (JakovTomasic) [inline client/src/components/Map.tsx:99]: You remove all old lines. That's great. You should remove all the old circles, too.

Comment (JakovTomasic) [reply client/src/components/Map.tsx:99]: After adding that line you will notice the duplicated code between the if and the else blocks.

Remove the else and extract the three lines below the if. The if block should only remove the image and redraw the new one.

File: client/src/components/Map.tsx
Comment (JakovTomasic) [inline client/src/components/Map.tsx:84]: Also, check if the flag is true (or check only that as if it's true its type wont be undefined).

There is a minor bug where someone can pass prop enableDrawNodes=false, but the drawing on click would still be enabled (false is not undefined).

File: client/src/components/SearchForm.tsx
Comment (JakovTomasic) [inline client/src/components/SearchForm.tsx:75]: Please just fix the indentation to improve readability. It's all over the place.

**Review by JakovTomasic — CHANGES_REQUESTED** [review]
Wow. This works great. It's really satisfying to click the next and back buttons 😆.

I requested a few changes mostly related to code quality rather than the core functionality.

The code is otherwise good, with well-separated segments (components, interfaces, functions).

You connected the home page with the navigation. Great job on that, too.

Can't wait for us to show this to everyone!

**Review by JakovTomasic — APPROVED** [review]
Great work! You may merge the pull request now 🥳

I assume there will be conflicts in package.json and yarn.lock files. In such a case, you can follow the instructions from the [Merge conflicts](https://github.com/JakovTomasic/RitehMaps/wiki/Solved-problems#merge-conflicts) section in the wiki.

Thank you for following all my change requests in all of the reviews. If you want me to better clarify the reasoning behind any of them, just ask. I hope you see the difference between now and the first commits.

Ask if you have any questions or prefer for me to merge the branch.

**Comment by JakovTomasic** [comment]
> This branch has no conflicts with the base branch

Ignore the before mentioned merge instructions and just press the big green button 😆

**Pull request merged**
